### PR TITLE
Redis Service

### PR DESCRIPTION
**Pull request by anatolysergeev**
add redis service that include codec and executor
all redis commands depend on redis service
remove redis executor dependency on codec

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -88,7 +88,7 @@ jobs:
       - name: Cache scala dependencies
         uses: coursier/cache-action@v6
       - name: Run test containers
-        run: docker-compose up -d
+        run: docker-compose -f testcontainers/redis-compose.yml up -d
```
Comment (mijicd) [inline .github/workflows/ci.yml:91]: This is not related to the change you're proposing so please revert it.

File: redis/src/main/scala/zio/redis/RedisCommand.scala [hunk 2]
```diff
@@ -16,17 +16,18 @@
...
 final class RedisCommand[-In, +Out] private (val name: String, val input: Input[In], val output: Output[Out]) {
-  private[redis] def run(in: In): ZIO[RedisExecutor, RedisError, Out] =
+  private[redis] def run(in: In): ZIO[Redis, RedisError, Out] =
     ZIO
-      .accessM[RedisExecutor] { executor =>
-        val service = executor.get
-        val codec   = service.codec
-        val command = Varargs(StringInput).encode(name.split(" "))(codec) ++ input.encode(in)(codec)
-        service.execute(command).flatMap[Any, Throwable, Out](out => ZIO.effect(output.unsafeDecode(out)(codec)))
+      .accessM[Redis] { redis =>
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RedisCommand.scala:25]: Let's replace `accessM` with `serviceWith`.

File: redis/src/main/scala/zio/redis/RedisUri.scala [hunk 3]
```diff
@@ -16,8 +16,8 @@
 
 package zio.redis
 
-final case class RedisConfig(host: String, port: Int)
+final case class RedisUri(host: String, port: Int)
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RedisUri.scala:19]: This is not related to the change you're proposing, please revert it.

File: redis/src/main/scala/zio/redis/package.scala [hunk 4]
```diff
@@ -39,6 +39,7 @@ package object redis
     with options.Scripting {
 
   type Id[+A]        = A
+  type Redis         = Has[Redis.Service]
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/package.scala:42]: Let's start moving towards the new module encoding. This assumes the following:
- remove the type aliases like this
- extract the `Service`s to top-level traits
- propagate

File: redis/src/test/scala/zio/redis/ScriptingSpec.scala [hunk 5]
```diff
@@ -24,7 +24,7 @@ trait ScriptingSpec extends BaseSpec {
                 |return redis.call('exists',KEYS[1])
               """.stripMargin
             res <- eval(lua, Chunk(key), Chunk(arg)).returning[Boolean]
-          } yield assertTrue(res)
+          } yield assert(res)(equalTo(true))
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/ScriptingSpec.scala:27]: This is not related to the change you're proposing, please revert it.

File: redis/src/main/scala/zio/redis/Redis.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+package zio.redis
+
+import zio.{Has, URLayer, ZIO, ZLayer}
+import zio.schema.codec.Codec
+
+object Redis {
+  trait Service {
+    def codec: Codec
+    def executor: RedisExecutor.Service
+  }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Redis.scala:26]: See the remark about new module structure given below.

File: redis/src/main/scala/zio/redis/Redis.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  private[this] final val RedisService =
+    ZLayer
+      .fromServicesManaged[RedisExecutor.Service, Codec, Any, Nothing, Redis.Service] { (e, c) =>
+        ZIO
+          .succeed(new Service {
+            override val codec: Codec                    = c
+            override val executor: RedisExecutor.Service = e
+          })
+          .toManaged_
+      }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Redis.scala:40]: This can be simplified:

```suggestion
    ZIO.services[Codec, RedisExecutor.Service].map { env =>
      new Service {
        val codec: Codec = c
        val executor: RedisExecutor.Service = e
      }
    }.toLayer
```

File: benchmarks/src/main/scala/zio/redis/BenchmarksUtils.scala [hunk 8]
```diff
@@ -17,22 +17,24 @@
...
-  def zioUnsafeRun(source: ZIO[RedisExecutor, RedisError, Unit]): Unit =
+  def zioUnsafeRun(source: ZIO[Redis, RedisError, Unit]): Unit =
     unsafeRun(source.provideLayer(BenchmarksUtils.Layer))
 }
 
 object BenchmarksUtils {
-  private final val Layer = Logging.ignore ++ ZLayer.succeed[Codec](StringUtf8Codec) >>> RedisExecutor.local.orDie
+  private final val Layer = {
+    val executor = Logging.ignore >>> RedisExecutor.local.orDie
+    executor ++ ZLayer.succeed[Codec](StringUtf8Codec) >>> Redis.live
```
Comment (mijicd) [inline benchmarks/src/main/scala/zio/redis/BenchmarksUtils.scala:38]: Just a heads up, once this PR lands, I'll submit the removal of `StringUtf8Codec`.

**Review by mijicd — CHANGES_REQUESTED** [review]

File: redis/src/main/scala/zio/redis/RedisUri.scala (hunk 3 above)
Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RedisUri.scala:19]: Can we use that name? 😇  I'am going to rename it in cluster PR anyway 😇

Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RedisUri.scala:19]: Have reverted

File: redis/src/main/scala/zio/redis/package.scala (hunk 4 above)
Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/package.scala]: Hope, I did it as you mean)

**Review by mijicd — APPROVED** [review]

File: redis/src/main/scala/zio/redis/RedisUri.scala (hunk 3 above)
Comment (mijicd) [reply redis/src/main/scala/zio/redis/RedisUri.scala]: I would still leave it :).
